### DISCUSSION TITLE
Can't install the thing - Installing Netbox RHEL 8

### DISCUSSION THREAD

**Discussion by fluffybunny2278** [netbox-community/netbox#8133] [Q&A]
Hello. I'm using RHEL 8 and am getting to the installation of Netbox. I cannot run the command from readthedocs page:

[root@ltech-dev1-netbox netbox-3.1.2]# sudo /opt/netbox/upgrade.sh
Removing old virtual environment...
Creating a new virtual environment at /opt/netbox-3.1.2/venv...
/opt/netbox/upgrade.sh: line 25: python3: command not found
--------------------------------------------------------------------
ERROR: Failed to create the virtual environment. Check that you have
the required system packages installed and the following path is
writable: /opt/netbox-3.1.2/venv

I try to create the venv folder manually but it looks like in this script, it will just remove it.

Anyone understand what is going on here?

**Comment by candlerb** [comment]
Apparently it can't find `python3`.

Note this step in the instructions [here](https://netbox.readthedocs.io/en/stable/installation/3-netbox/#install-system-packages):

> Before continuing, check that your installed Python version is at least 3.7:
> 
> `python3 -V`

Was that command successful?  If yes, then please show the output of `which python3`.  Also, did you install python3 from somewhere other than the default RHEL repositories?

> I try to create the venv folder manually but it looks like in this script, it will just remove it.

Indeed it will, but it will also recreate it, so it needs to have permissions to do so.  If you're running `upgrade.sh` as root, then this won't be an issue.

**Reply by fluffybunny2278** [reply]
[root@ltech-dev1-netbox netbox-3.1.2]# python3.9 --version
Python 3.9.9

ohhhhhhhhhh.... maybe I have to update the script to say "python3.9" and not "python3"?


I installed python using this method which I believe is NOT from the RHEL repositories:
(https://tecadmin.net/install-python-3-7-on-centos-8/)

cd /opt
wget https://www.python.org/ftp/python/3.7.9/Python-3.7.9.tgz

**Reply by fluffybunny2278** [reply]
I realized that I did not do what you asked.... 

[netman@ltech-dev1-netbox netbox]$ python3 -V
-bash: python3: command not found
[netman@ltech-dev1-netbox netbox]$ which python3
/usr/bin/which: no python3 in (/home/netman/.local/bin:/home/netman/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin)
[netman@ltech-dev1-netbox netbox]$

**Reply by fluffybunny2278** [reply]
Here is me installing from the repositories. python3 -V is still not found as a valid command.

[netman@ltech-dev1-netbox netbox]$ sudo yum install -y gcc libxml2-devel libxslt-devel libffi-devel libpq-devel openssl-devel redhat-rpm-config
[sudo] password for netman: 
Updating Subscription Management repositories.
Last metadata expiration check: 2:36:42 ago on Tue 21 Dec 2021 07:40:23 AM MST.
Package gcc-8.5.0-4.el8_5.x86_64 is already installed.
Package libxml2-devel-2.9.7-9.el8_4.2.x86_64 is already installed.
Package libxslt-devel-1.1.32-6.el8.x86_64 is already installed.
Package libffi-devel-3.1-22.el8.x86_64 is already installed.
Package libpq-devel-13.3-1.el8_4.x86_64 is already installed.
Package openssl-devel-1:1.1.1k-5.el8_5.x86_64 is already installed.
Package redhat-rpm-config-125-1.el8.noarch is already installed.
Dependencies resolved.
Nothing to do.
Complete!

[netman@ltech-dev1-netbox netbox]$ python3 -V
-bash: python3: command not found
[netman@ltech-dev1-netbox netbox]$

**Reply by candlerb** [reply]
Hold your horses.  All you need is:

```
PYTHON=/usr/bin/python3.9 /opt/netbox/upgrade.sh
```

as described [here](https://netbox.readthedocs.io/en/stable/installation/3-netbox/#run-the-upgrade-script).

(Note that you'll also have to remember to do that each time you upgrade Netbox in future)

**Reply by fluffybunny2278** [reply]
whooaaaaa nelly! :D 

Okay, long story short... I think I got it! I didn't use that command you listed because I read the text as only use that if it's version 3.7 or below. After running your command, I got a pip error. Then I figured out how to install stupid pip and now that command is working. I'm going to go back to my snapshot and do this all from the beginning.... I've entered so many commands it's probably a mess. Standby!

**Reply by fluffybunny2278** [reply]
I found where this Python thing is supposed to be:
[netman@ltech-dev1-netbox Python-3.9.9]$ PYTHON=/usr/bin/python3 /opt/netbox/upgrade.sh  
Creating a new virtual environment at /opt/netbox-3.1.2/venv...

So now it's telling me that I have Python 3.6.8 installed, no... I don't. I have 3.9.9. I probably have 2 versionf of Python installed in this stupid thing somehow.

Then I looked more and found a new directory:
sudo PYTHON=/opt/Python-3.9.9/python /opt/netbox/upgrade.sh

Anyway..... 

I would rather step in front of a moving train while being eaten alive by piranhas naked covered in something that makes bees sting and smash a beehive on the ground in front of me than have to use Linux again. All that being said, this crap is finally working I think. I installed it twice and got it to work. 

Thanks so much @candlerb  for all your pateience and helping me understand! I do really appreciate you!

Hopefully I can install this thing in production now with no issues. Make sure to leave your cell number so when it breaks, I can call you... hahah :D I'm now going to make a document as I install the final time in prod of how to do all this crap so if/when it has to be rebuilt, it will be easy for me or whoever else has to do this. Thanks again!!

**Reply by candlerb** [reply]
Maybe you could check whether [Netbox Cloud](https://ns1.com/blog/introducing-netbox-cloud-by-ns1) is still a thing.

**Comment by DanSheps** [comment]
Off topic, but...  

Did anyone think of this with that topic?  https://i.kym-cdn.com/photos/images/original/000/834/323/bdb.gif

**Comment by fluffybunny2278** [comment]
I wanted to post my text file to help anyone else who may be dumb at Linux like I am, perhaps it will help someone in the future. My next steps are to get TLS working which I'll need to get a cert from our systems team, get LDAP auth working (more systems team), and then I'm trying to create an alias to CD and then run python3 + arguments. Right now, no workie. This is what I have so far, if anyone knows exactly why it doesn't work, I just started looking into it today.

**alias netbox='( cd "/opt/netbox/netbox" && python3 manage.py runserver 0.0.0.0:6969 --insecure )'**
netbox
-bash: python3: command not found


[Netbox install RHEL 8.txt](https://github.com/netbox-community/netbox/files/7802568/Netbox.install.RHEL.8.txt)

**Reply by candlerb** [reply]
> This is what I have so far, if anyone knows exactly why it doesn't work, I just started looking into it today.
> 
> **alias netbox='( cd "/opt/netbox/netbox" && python3 manage.py runserver 0.0.0.0:6969 --insecure )'**

You Don't Want To Do That™.  The build-in "runserver" is [for development purposes only](https://netbox.readthedocs.io/en/stable/installation/3-netbox/#test-the-application); it is neither robust nor secure enough for production use.

What you need to do is to run netbox [gunicorn](https://netbox.readthedocs.io/en/stable/installation/4-gunicorn/), which is managed as a service under systemd (hence you don't need a startup alias).  There is also a separate systemd service for netbox-rq, which is the internal queueing service used by Netbox for things like dispatching webhooks and running custom scripts.  Then you have a [proxy front-end](https://netbox.readthedocs.io/en/stable/installation/5-http-server/) (Apache or Nginx), which is also a systemd service. You will in any case have to do this in order to get TLS working, since runserver doesn't do that.

This is all in the documentation; if you have a friendly systems team then they should be able to help you with it.

But for completeness, you should be able to get runserver to work as follows:

```
cd /opt/netbox
source venv/bin/activate
python netbox/manage.py runserver 0.0.0.0:6969 --insecure
```

Note "python" not "python3" there.  It will pick up the "python" within the virtualenv, which will be whatever version of python the venv was built with (in your case `/opt/Python-3.9.9/python`)

> This takes you into the virtual environment that NetBox is running in... don't ask me how to get out of it???

`deactivate`

**Reply by fluffybunny2278** [reply]
candlerb, you're a certified genius! Thank you so much for all your help, I'll take a look at this gunicorn thing.

I like the "You Don't Want To Do That™"   :D   At least I can still laugh with all this Linux frustration.

**Reply by fluffybunny2278** [reply]
@candlerb ..... sooooooooo Do you know anything about nginx?

I was able to install the Gunicorn with no issues. However, I need to create a webserver now I guess. nginx seems to be easier so I'm going to go with that, I did try Apache too and couldn't get it working. I'm certain that it has something to do with file structure again since the Netbox docs say to use deprecated commands/directories. 

Here is the error I have:
[netman@ltech-dev1-netbox conf.d]$ sudo nginx -t
nginx: [emerg] "server" directive is not allowed here in **/etc/nginx/default.d/netbox.conf:1**
nginx: configuration file /etc/nginx/nginx.conf test failed

This is the netbox.conf file:
server {
    listen 80;
    server_name 10.2.0.205;
    client_max_body_size 25m;

    location /static/ {
        alias /opt/netbox/netbox/static/;
    }

    location / {
        proxy_pass http://127.0.0.1:8001;
    }
}




Which I am assuming is referenced by the  nginx.conf file:
```# For more information on configuration, see:
```#   * Official English Documentation: http://nginx.org/en/docs/
```#   * Official Russian Documentation: http://nginx.org/ru/docs/

user nginx;
worker_processes auto;
error_log /var/log/nginx/error.log;
pid /run/nginx.pid;

# Load dynamic modules. See /usr/share/doc/nginx/README.dynamic.
include /usr/share/nginx/modules/*.conf;

events {
    worker_connections 1024;
}

http {
    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log  /var/log/nginx/access.log  main;

    sendfile            on;
    tcp_nopush          on;
    tcp_nodelay         on;
    keepalive_timeout   65;
    types_hash_max_size 2048;

    include             /etc/nginx/mime.types;
    default_type        application/octet-stream;

    # Load modular configuration files from the /etc/nginx/conf.d directory.
    # See http://nginx.org/en/docs/ngx_core_module.html#include
    # for more information.
    include /etc/nginx/conf.d/*.conf;

    server {
        listen       80 default_server;
        listen       [::]:80 default_server;
        server_name  _;
        root         /usr/share/nginx/html;

        **# Load configuration files for the default server block.
        include /etc/nginx/default.d/*.conf;**

        location / {
        }

        error_page 404 /404.html;
            location = /40x.html {
        }

        error_page 500 502 503 504 /50x.html;
            location = /50x.html {
        }
    }




Now, It seems that I can comment out the INCLUDE and then change the LISTEN parameter to 10.2.0.205:80 and then I successfully get the test nginx page when I navigate. However, I still don't load Netbox, only the test page for nginx.

**Comment by candlerb** [comment]
I don't use nginx, I use apache.

However in any case I think you'd be better served by your local sysadmins for help with things like this.

BTW, your post is mangled somewhat because of the markdown rendering in github.  Please edit your post and put three backticks (` ``` `) on a line of their own, before and the contents of each file that you posted.
